blog/2020/05/invalidations.md

```diff
@@ -0,0 +1,518 @@
+
+julia> f(::Missing) = 5
+f (generic function with 5 methods)
+
+julia> @btime applyf($c)
+  33.537 ns (0 allocations: 0 bytes)
+3
+```
+It's almost a tenfold difference.
+If `applyf` is performance-critical, you'll be very happy that Julia tries to give you the best version it can, given the available information.
+But this leaves the door open to invalidation, which means recompilation the next time you use `applyf`.
+If method invalidation happens often, this might contribute to making Julia "feel" sluggish.
+
+# How common is method invalidation?
+
+Unfortunately, method invalidation is pretty common.
+First, let's get some baseline statistics.
+Using the [MethodAnalysis] package (which is at a very early stage of development
+at the time of this writing), you can find out that a fresh Julia session
+(albeit one that has loaded the MethodAnalysis package and used it to perform some analysis) has almost 50,000 `MethodInstance`s tucked away in its cache.
+These are mostly for `Base` and the standard libraries.
+
+Using some not-yet merged work in both Julia itself and [SnoopCompile], we can count the number of invalidations when we load various packages into a fresh Julia session:
+
+| Package | # of unique invalidations |
+|:------- | ------------------:|
+| Example | 0 |
+| Revise | 27 |
+| FixedPointNumbers | 429 |
+| SIMD | 2799 |
+| StaticArrays | 2852 |
+| Optim | 3171 |
+| Images | 3638 |
+| Flux | 3697 |
+| Plots | 4002 |
+| DataFrames | 4048 |
+| JuMP | 4666 |
+| Makie | 6118 |
+| DifferentialEquations | 6777 |
+
+You can see that key packages used by large portions of the Julia ecosystem invalidate
+hundreds or thousands of MethodInstances, sometimes more than 10% of the total
+number of MethodInstances present before loading the package.
+
+# How serious is method invalidation?
+
+The next time you want to call functionality that gets invalidated,
+you have to wait for recompilation.
+We can illustrate this using everyone's favorite example, plotting:
+
+```julia
+julia> using Plots
+
+julia> @time display(plot(rand(5)))
+  7.717729 seconds (15.27 M allocations: 797.207 MiB, 3.59% gc time)
+```
+
+As is well known, it's much faster the second time, because it's already compiled:
+
+```julia
+julia> @time display(plot(rand(5)))
+  0.311226 seconds (19.93 k allocations: 775.055 KiB)
+```
+
+Now load a package that does a lot of invalidation, and try again:
+
+```julia
+julia> using SIMD
+
+julia> @time display(plot(rand(5)))
+  7.238336 seconds (26.50 M allocations: 1.338 GiB, 7.88% gc time)
+```
+
+Because so much got invalidated by loading SIMD, Julia had to recompile many methods before it could once again produce a plot, so that in terms of time it was almost as expensive as the first usage.
+The size of the effect varies substantially depending on what task you are trying to achieve and what packages you load to do the invalidation.
+(If you do this with StaticArrays rather than SIMD there's no increase in latency,
```

Comment:
Maybe demonstrate this first. I feel like the reader might be getting here like "oh no, Julia is doomed!" so it might be good to start by saying "loading some packages doesn't effect ... . Here StaticArrays ... however notice SIMD ..."